File: C++/726. Number of Atoms.cpp
class Solution {
public:
    bool isDigit(char ch) {
        return (ch >= '0' && ch <= '9');
    }    
    unordered_map<string, int> dfs(int &pos, string &formula) {
        int n = formula.size();
        if (pos >= n) {
            return {};
        }
        unordered_map<string, int> curMap;
        string prev = "";
        while (pos < n) {
            if (formula[pos] == '(') {
                if (prev.size() > 0) {
                    curMap[prev] += 1;
                    prev = "";
                }
                ++pos;
                unordered_map<string, int> retMap = dfs(pos, formula);
                for (pair<string, int> entry: retMap) {
                    curMap[entry.first] += entry.second;   
                }
            } else if (formula[pos] == ')') {
                if (prev.size() > 0) {
                    curMap[prev] += 1;
                    prev = "";
                }
                int val = 0;
                if (pos + 1 < n) {
                    ++pos;
                    while (pos < n) {
                        if (isDigit(formula[pos])) {
                            val = val * 10 + formula[pos] - '0';
                        } else break;
                        ++pos;
                    }
                }
                val = val > 0? val : 1;
                for (auto &entry: curMap) {
                    entry.second *= val;
                }
                return curMap;
            } else {
                if (isDigit(formula[pos])) {
                    int val = 0;
                    while (pos < n) {
                        if (isDigit(formula[pos]))
                            val = val * 10 + formula[pos] - '0';
                        else break;
                        ++pos;
                    }
                    if (prev.size() > 0) {
                        curMap[prev] += val;
                        prev = "";
                    }                
                } else {
                    if (prev.size() > 0 && isupper(formula[pos])) {
                        curMap[prev] += 1;
                        prev = "";
                    }
                    prev += formula[pos];
                    ++pos;
                }
            }
        }
        if (prev.size() > 0) {
            curMap[prev] += 1;
        }
        return curMap;
    }
    
    string countOfAtoms(string formula) {
        int pos = 0;
        unordered_map<string, int> f = dfs(pos, formula);
        vector<string> res;
        for (pair<string, int> entry: f) {
            res.push_back(entry.first);
        }
        sort(begin(res), end(res), [&] (const string &a, const string &b) {
            return a < b;
        });
        string result = "";
        for (string s: res) {
            result += s;
            if (f[s] > 1)
                result += to_string(f[s]);
        }
        return result;
    }
};

File: JAVA/726. Number of Atoms.java
import java.util.Map.Entry;
class Solution {
     private int index = 0;
    public String countOfAtoms(final String formula) {
        StringBuilder stringBuilder = new StringBuilder();
        for (Entry<String, Integer> entry : makeChemicalFormula(formula).entrySet()) {
            String key = entry.getKey();
            int value = entry.getValue();
            stringBuilder.append(key);
            if (value != 1) {
                stringBuilder.append(value);
            }
        }
        return stringBuilder.toString();
    }
    private Map<String, Integer> makeChemicalFormula(final String formula) {
        Map<String, Integer> map = new TreeMap<>();
        while (index < formula.length()) {
            char current = formula.charAt(index);
            if (current == '(') {
                index++;
                for (Entry<String, Integer> entry : makeChemicalFormula(formula).entrySet()) {
                    String key = entry.getKey();
                    int value = entry.getValue();
                    map.merge(key, value, Integer::sum);
                }
            } else if (current == ')') {
                index++;
                int num = getNum(formula);
                for (String key : map.keySet()) {
                    int value = map.get(key);
                    map.put(key, value * num);
                }
                return map;
            } else {
                String atom = getAtom(formula);
                int num = getNum(formula);
                map.merge(atom, num, Integer::sum);
            }
        }
        return map;
    }
    private String getAtom(final String formula) {
        int startIndex = index++;
        int n = formula.length();
        while (index < n && Character.isLowerCase(formula.charAt(index))) {
            index++;
        }
        return formula.substring(startIndex, index);
    }
    private int getNum(final String formula) {
        int startIndex = index;
        int n = formula.length();
        while (index < n && Character.isDigit(formula.charAt(index))) {
            index++;
        }
        String numString = formula.substring(startIndex, index);
        return numString.isEmpty() ? 1 : Integer.parseInt(numString);
    }
}
